src/util/battery/batterylinux.cpp

```diff
@@ -7,99 +7,61 @@
 #include <QtDebug>
 
 BatteryLinux::BatteryLinux(QObject* pParent)
-    : Battery(pParent) {
+    : Battery(pParent),
+      m_client(up_client_new()) {
 }
 
 BatteryLinux::~BatteryLinux() {
+    g_object_unref(static_cast<UpClient*>(m_client));
 }
 
 void BatteryLinux::read() {
-    m_iMinutesLeft = Battery::TIME_UNKNOWN;
-    m_dPercentage = 0.0;
-    m_chargingState = Battery::UNKNOWN;
-
-    // NOTE(rryan): It would be nice if we could create the client
-    // once. However, while testing this up_client_get_devices(client) returned
-    // an empty list when I tried to re-use the UpClient instance.
-    UpClient* client = up_client_new();
-    VERIFY_OR_DEBUG_ASSERT(client) {
-      return;
+    VERIFY_OR_DEBUG_ASSERT(static_cast<UpClient*>(m_client)) {
+        return;
     }
 
-#if !UP_CHECK_VERSION(0, 9, 99)
-    // Re-enumerate in case a device is added.
-    up_client_enumerate_devices_sync(client, NULL, NULL);
-#endif
-
-#if UP_CHECK_VERSION(0, 99, 8)
-    GPtrArray* devices = up_client_get_devices2(client);
-    VERIFY_OR_DEBUG_ASSERT(devices) {
-      return;
+    guint kind;
+    gboolean isPresent;
+    guint state;
+    gdouble percentage;
+    gint64 timeToEmpty;
+    gint64 timeToFull;
+    UpDevice *device = up_client_get_display_device(static_cast<UpClient*>(m_client));
+    if (device) {
```

Comment:
I didn't notice that `isPresent` was also a parameter.

But wouldn't the code get simple by using an early exit strategy?

```
if (!device || !isPresent) {
    return;
}
...read battery status...
```

I'm not a fan of deeply nested if blocks, especially if the if/else branches are asymmetric like in this case. A linear control flow is often easier to read.

Reply:
> if (!device || !isPresent) {
>     return;
> }
> ...read battery status...

This doesn't work because the battery status has to be read first to get the `isPresent` value. 
Anyway, I have now pushed a version doing `early exit` where it is possible.